### ISSUE TITLE
LTRdigest stuck for ten days

### ISSUE THREAD

**Issue by lhui2010** [genometools/genometools#970]
## Problem description

Hi, 

I was running LTRdigest on the result of LTRharvest. The process is still running (in S status). But the output stopped about ten days ago. I used strace -p to see what's going on for gt and hmmscan. The output is not not updating. 

gt
```
strace: Process 91896 attached
write(13, "\n", 1
```

hmmscan
```
$strace -p 333397 
strace: Process 333397 attached
write(1, "\n  == domain 2  score: 73.5 bits"..., 523
```

$lt output
```
  0 May 14 17:53 Sp.genome.ltrdigest_3ltr.fas
  0 May 14 17:53 Sp.genome.ltrdigest_5ltr.fas
  0 May 14 17:53 Sp.genome.ltrdigest_complete.fas
23K May 14 17:53 Sp.genome.ltrdigest_conditions.csv
  0 May 14 17:53 Sp.genome.ltrdigest_pbs.fas
  0 May 14 17:53 Sp.genome.ltrdigest_ppt.fas
  0 May 14 17:53 Sp.genome.ltrdigest_tabout.csv
```

## Exact command line call triggering the problem

```
gt -j 10 ltrdigest -pptlen 10 30 -pbsoffset 0 3 -trnas /lustre/local/database/tRNA/eukaryotic-tRNAs.fa -hmms /lustre/local/database/gydb/GyDB_collection/profiles/AP_1731.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_17_6.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_412_mdg1.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_a_clade.hmm/lustre/local/database/gydb/GyDB_collection/profiles/AP_alpharetroviridae.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_athila.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_badnavirus.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_b_clade.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_bel.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_betaretroviridae.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_caulimoviridae_dom2.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_caulimovirus.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_cavemovirus.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_cer2_3.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_codi_c.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_codi_d.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_codi_I.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_codi_II.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_cog3577.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_cog5550.hmm /lustre/local/database/gydb/GyDB_collection/profiles/AP_copia.hmm /lustre/local/database/gydb/GyDB_collection/ ...
-outfileprefix ltrdigest/Sp.genome.ltrdigest Sp.genome.ltrharvest.sorted.gff3 Sp.genome

```

## Example minimal input triggering the problem
```
head Sp.genome.ltrharvest.sorted.gff3 
==> Sp.genome.ltrharvest.sorted.gff3 <==
##gff-version 3
##sequence-region   seq0 1 33263194
##sequence-region   seq10 1 44760251
##sequence-region   seq2 1 81372794
##sequence-region   seq3 1 56792611
##sequence-region   seq4 1 56999476
##sequence-region   seq5 1 53814584
##sequence-region   seq6 1 61110319
##sequence-region   seq7 1 39818848
##sequence-region   seq8 1 41393046
##sequence-region   seq9 1 37631608
#10
#1
#3
#2
#5
#4
#7
#6
#9
#8
seq0    LTRharvest      repeat_region   8456236 8553983 .       ?       .       ID=repeat_region1
seq0    LTRharvest      target_site_duplication 8456236 8456240 .       ?       .       Parent=repeat_region1
seq0    LTRharvest      inverted_repeat 8456241 8456242 .       ?       .       Parent=repeat_region1
seq0    LTRharvest      LTR_retrotransposon     8456241 8553978 .       ?       .       ID=LTR_retrotransposon1;Parent=repeat_region1;ltr_similarity=88.
seq0    LTRharvest      long_terminal_repeat    8456241 8458126 .       ?       .       Parent=LTR_retrotransposon1
seq0    LTRharvest      long_terminal_repeat    8552067 8553978 .       ?       .       Parent=LTR_retrotransposon1
seq0    LTRharvest      inverted_repeat 8458125 8458126 .       ?       .       Parent=repeat_region1

```

## What GenomeTools version are you reporting an issue for (as output by `gt -version`)?
1.5.10

## Did you compile GenomeTools from source? If so, please state the `make` parameters used.
I used conda to install gt.

## What operating system (e.g. Ubuntu, Mac OS X), OS version (e.g. 15.10, 10.11) and platform (e.g. x86_64) are you using?
CentOS7, x86_64

**Comment by satta** [comment]
I does not look like you're doing something wrong in principle.
How many candidates do you have in your input GFF? Could you maybe retry your run with a smaller set (maybe even a single candidate) to verify if it terminates at all? Or does your snippet above actually mean you only have this one candidate?

Otherwise, there's not much I can do without the full (or representative, triggering the issue) set of data to reproduce.

**Comment by lhui2010** [comment]
Hi, 

There were only 35 candidant regions in my GFF. But the problem persist even if I used only one TE annotation (ltrdigest becomes sleeping status and not exiting). 

Previously I also used similar command and there were no such problem. I have uploaded the problematic fasta and sorted gff files below. Can you have a look? Thanks!

[Sp_test_run.tar.gz](https://1drv.ms/u/s!AmpVypLR8UBBkohWqs-LRoGs0Ic67Q?e=eCgxdm)

Hui

**Comment by satta** [comment]
I just managed to find some time to get a look. I used the latest versions of the GyDB pHMMs and GtRNADB for that.

First of all, I can confirm your observation. LTRdigest gets stuck with no CPU load. I noticed that this only happens when you specify all pHMM files from GyDB at once.  Just out of curiosity, I just picked one of the  -- worked. Used `profiles/AP_*` -- worked, with proper `protein_match`es in the GFF.

My suspicion is that either one of the HMMs is the problem, or the size of the model set may be relevant. I am puzzled, because I remember running LTRdigest with all GyDB models before with no problem. I think it might be helpful playing with different variations of the models and try to narrow down the problem. I will also do some debugging to see what's happening later next week.

**Comment by satta** [comment]
I think I have some idea of where the problem is now. LTRdigest is using pipe buffers to first write the input sequences (6-frame translation of the retrotransposon sequence) into `hmmscan`'s input, and only starts to read after all input is sent. However, since `hmmscan` is starting to send data while already having received data from a buffer flush, LTRdigest is not reading yet from `hmmscan`'s output, hence blocking the data flow.

I was able to work around this by increasing the pipe buffer size via `fcntl()` to 512KB (from the default 64KB) in the code, but that is IMHO not a proper solution since it is not very portable and basically just a workaround. The proper way to do this is by doing reading and writing from/to `hmmscan` in separate threads. I am going to look into this, but can't guarantee any finished date so far, sorry.

**Comment by satta** [comment]
I have addressed this in #971.

**Comment by lhui2010** [comment]
Hi, 

Thanks for all your time and efforts! It's great to read your response!

Sorry I was traveling recently and logged on github. On my side, I found a temporary solution by switching to genome tools 1.5.1. Also, when using the lower version of gt, I found one hmm name of GyDB contained slash and causing error. I manually renamed this hmm. But this doesn't look like the bug causing all the problems.

I'm downloading the newest genometools and will have a try with my data. Thanks for the fix! :)

Hui
